### PR TITLE
[Enhancement] Support split_part function negative index

### LINKED ISSUE #5992: Function split_part  not support "-1" index
### PostgreSQL
middle_img_v2_2eda0ea8-9621-4805-8477-f826843b977g
### StarRocks
middle_img_v2_c3c12e22-18d5-4a41-b6b2-6a428c51e39g

### PR DESCRIPTION
**Pull request by leoyy0316**
Fixes #5992

### CODE REVIEW

**Review by imay — COMMENTED** [review]
needs to add more tests in sql-test

File: be/src/exprs/split_part.cpp [hunk 1]
```diff
@@ -78,55 +73,89 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
...
-                char* pos = reinterpret_cast<char*>(memchr(haystack.data + offset + 1, delimiter.data[0], n));
-                if (pos != nullptr) {
-                    offset = pos - haystack.data;
-                    num++;
-                } else {
-                    offset = haystack.size;
-                    num = (num == 0) ? 0 : num + 1;
-                    break;
+        }
+        if (part_number > 0) {
```
Comment (LiShuMing) [inline be/src/exprs/split_part.cpp:77]: It's weird here, add else ?

File: be/src/exprs/split_part.cpp [hunk 2]
```diff
@@ -78,55 +73,89 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
...
+
+                if (num == part_number) {
+                    res.append(
+                            Slice(haystack.data + pre_offset + delimiter.size, offset - pre_offset - delimiter.size));
+                } else {
+                    res.append_null();
+                }
             }
         } else {
-            // if delimiter is a string, use memmem to split
```
Comment (LiShuMing) [inline be/src/exprs/split_part.cpp:109]: Can the positive matching (where part_number is positive) and negative matching be extracted and encapsulated separately? This way, the original logic can be preserved without any impact?

File: be/src/exprs/split_part.cpp [hunk 3]
```diff
@@ -78,55 +73,89 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
...
-                pre_offset = offset;
-                size_t n = haystack.size - offset - delimiter.size;
-                char* pos = reinterpret_cast<char*>(
-                        memmem(haystack.data + offset + delimiter.size, n, delimiter.data, delimiter.size));
-                if (pos != nullptr) {
-                    offset = pos - haystack.data;
-                    num++;
+            auto substr = haystack_str;
+            while (num <= part_number && offset >= 0) {
+                offset = (int)substr.rfind(delimiter, offset);
```
Comment (LiShuMing) [inline be/src/exprs/split_part.cpp:140]: Could this implementation cause performance issues? Can it be implemented using memmem for reverse matching?

File: be/src/exprs/split_part.cpp [hunk 4]
```diff
@@ -24,6 +24,96 @@
 
 namespace starrocks {
 
+static void split_positive_index(ColumnBuilder<TYPE_VARCHAR> res, Slice& delimiter, int32_t& part_number,
```
Comment (LiShuMing) [inline be/src/exprs/split_part.cpp:27]: to be simpler, no need to add `ColumnBuilder<TYPE_VARCHAR> res` as a parameter? how about 
`static Slice split_positive_index( Slice& haystack, Slice& delimiter, int32_t part_number)` instead?

**Review by LiShuMing — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]
LGTM

**Review by Seaven — DISMISSED** [review]
LGTM

File: be/src/exprs/split_part.cpp [hunk 5]
```diff
@@ -24,6 +24,96 @@
...
+
+        if (num == part_number) {
+            res.append(Slice(haystack.data + pre_offset + delimiter.size, offset - pre_offset - delimiter.size));
+        } else {
+            res.append_null();
+        }
+    }
+}
+
+static void split_negative_index(ColumnBuilder<TYPE_VARCHAR> res, Slice& delimiter, int32_t& part_number,
```
Comment (satanson) [inline be/src/exprs/split_part.cpp:84]: split_negative_index and split_positive_index seems similar, cound we refactor these two functions into a function template and use template parameter to handle the different parts.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exprs/split_part.cpp [hunk 6]
```diff
@@ -24,6 +24,94 @@
...
+                    num = (num == 0) ? 0 : num + 1;
+                    break;
+                }
+            }
+
+            if (num == part_number) {
+                res = Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1);
+            } else {
+                res;
+            }
```
Comment (satanson) [inline be/src/exprs/split_part.cpp:55]: change

```cpp
           if (num == part_number) {
                res = Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1);
            } else {
                res;
            }
```

to

```cpp
           if (num == part_number) {
                res = Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1);
            } 
```

File: be/src/exprs/split_part.cpp [hunk 7]
```diff
@@ -24,6 +24,94 @@
...
+                    offset = pos - haystack.data;
+                    num++;
+                } else {
+                    offset = haystack.size;
+                    num = (num == 0) ? 0 : num + 1;
+                    break;
+                }
+            }
+
+            if (num == part_number) {
```
Comment (satanson) [inline be/src/exprs/split_part.cpp:76]: change

```cpp
           if (num == part_number) {
                res = Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1);
            } else {
                res;
            }
```

to

```cpp
           if (num == part_number) {
                res = Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1);
            } 
```

File: be/src/exprs/split_part.cpp [hunk 8]
```diff
@@ -24,6 +24,94 @@
...
+        }
+    } else {
+        part_number = -part_number;
+        auto haystack_str = haystack.to_string();
+        int32_t offset = haystack.size;
+        int32_t pre_offset = offset;
+        int32_t num = 0;
+        auto substr = haystack_str;
+        while (num <= part_number && offset >= 0) {
+            offset = (int)substr.rfind(delimiter, offset);
```
Comment (satanson) [inline be/src/exprs/split_part.cpp:90]: When delimiter.size()=1, rfind and memrchr,  which are better?  
Please leave a TODO mark to notice  benchmarking rfind vs memrchr in the future.

File: be/src/exprs/split_part.cpp [hunk 9]
```diff
@@ -34,9 +122,9 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
     RETURN_IF_COLUMNS_ONLY_NULL(columns);
 
     if (columns[2]->is_constant()) {
```
Comment (satanson) [inline be/src/exprs/split_part.cpp:124]: it seems there exists a SIMD algorithm for split_part when delimiter is constant, the sketch of the algo as follows:
1. search the delimiter BinaryColumn::bytes() and save the offsets of occurrences in a array.
2. then compute the split_port from BinaryColumn' offsets and the offsets from step 1.

We can leave a TODO to notice somebody for future optimization.

**Review by satanson — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**